File: src/main/java/com/morak/performancetracker/context/ResultMapper.java
package com.morak.performancetracker.context;

import com.morak.performancetracker.Monitor;
import com.morak.performancetracker.aop.persistence.QueryMonitor;
import com.morak.performancetracker.aop.web.WebMonitor;
import org.springframework.stereotype.Component;

@Component
public class ResultMapper {

    public Result mapMonitor(Monitor monitor) {
        Result result = mapToResult(monitor);
        monitor.clear();
        return result;
    }

    private Result mapToResult(Monitor monitor) {
        if (monitor instanceof QueryMonitor) {
            QueryMonitor qm = (QueryMonitor) monitor;
            return new Result(qm.getQuery(), qm.getQueryTime());
        }
        if (monitor instanceof WebMonitor) {
            WebMonitor wm = (WebMonitor) monitor;
            return new Result(wm.getMethod() + " " + wm.getUri(), wm.getElapsed());
        }
        throw new IllegalArgumentException("Couldn't find proper monitor");
    }
}

File: src/main/kotlin/com/morak/performancetracker/context/ResultMapper.kt
package com.morak.performancetracker.context

import com.morak.performancetracker.Monitor
import com.morak.performancetracker.aop.persistence.QueryMonitor
import com.morak.performancetracker.aop.rest.RestMonitor
import com.morak.performancetracker.aop.web.WebMonitor
import org.springframework.stereotype.Component

@Component
class ResultMapper {
    fun mapMonitor(monitor: Monitor): MonitorResult {
        val result = mapToResult(monitor)
        monitor.clear()
        return result
    }

    private fun mapToResult(monitor: Monitor): MonitorResult {
        return when (monitor) {
            // todo: remove "!!"
            is QueryMonitor -> MonitorResult(monitor.query!!, monitor.queryTime)
            is WebMonitor -> MonitorResult(monitor.method + " " + monitor.uri, monitor.elapsed.toDouble())
            is RestMonitor -> MonitorResult(monitor.uri!!, monitor.elapsedTime.toDouble())
            else -> throw IllegalArgumentException("Couldn't find proper monitor")
        }
    }
}
